File: src/main/java/ru/arc/leafdecay/LeafDecayManager.java
package ru.arc.leafdecay;

import com.destroystokyo.paper.ParticleBuilder;
import com.jeff_media.customblockdata.CustomBlockData;
import lombok.extern.slf4j.Slf4j;
import org.bukkit.*;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;
import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.util.ParticleManager;

import java.util.*;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.ConcurrentSkipListSet;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;


@Slf4j
public class LeafDecayManager {

    public static LeafChecker leafChecker;
    private static BukkitTask checkTask;
    private static BukkitTask decayTask;
    private static BukkitTask tagClearTask;

    private static final ConcurrentSkipListSet<Chunk> chunkQueue =
            new ConcurrentSkipListSet<>(Comparator.comparingInt(Chunk::getX).thenComparingInt(Chunk::getZ));
    private static final Deque<Location> leafQueue = new ConcurrentLinkedDeque<>();
    private static final Deque<Block> awaitingTagClearing = new ConcurrentLinkedDeque<>();
    private static final NamespacedKey playerPlacedKey = new NamespacedKey(ARC.plugin, "lf");
    private static final Config config = ConfigManager.of(ARC.plugin.getDataPath(), "leafdecay.yml");

    private static long decayInterval = 5L;
    private static long checkInterval = 20L;
    private static long checkWorldEach = 1000L;
    private static long leafBatchSize = 100L;
    private static Set<Material> leafMaterials;
    private static Set<Material> trunkMaterials;
    private static boolean playParticles = true;
    private static boolean playSound = true;
    private static Set<String> worlds = Set.of();

    public static void init() {
        reload();
    }

    public static void reload() {
        decayInterval = config.integer("decay-interval", 5);
        checkInterval = config.integer("global-check-interval", 20);
        checkWorldEach = config.integer("world-check-interval", 1000);
        leafBatchSize = config.integer("leaf-batch-size", 100);
        leafMaterials = config.<String>list("leaf-materials").stream()
                .map(String::toUpperCase)
                .map(Material::valueOf)
                .collect(Collectors.toSet());
        trunkMaterials = config.<String>list("trunk-materials").stream()
                .map(String::toUpperCase)
                .map(Material::valueOf)
                .collect(Collectors.toSet());
        leafChecker = new LeafChecker(config, leafMaterials, trunkMaterials);
        playParticles = config.bool("play-particles", true);
        playSound = config.bool("play-sound", true);
        worlds = new HashSet<>(config.stringList("leaf-decay-worlds"));

        start();
    }

    public static void cancel() {
        if (checkTask != null && !checkTask.isCancelled()) checkTask.cancel();
        if (decayTask != null && !decayTask.isCancelled()) decayTask.cancel();
        if (tagClearTask != null && !tagClearTask.isCancelled()) tagClearTask.cancel();
    }

    public static void start() {
        cancel();
        AtomicInteger counter = new AtomicInteger();
        checkTask = new BukkitRunnable() {
            @Override
            public void run() {
                int count2 = 0;
                if (leafQueue.size() > 10000) return;
                while (!chunkQueue.isEmpty()) {
                    if (count2++ > 5) break;
                    Chunk chunk = chunkQueue.pollFirst();
                    if (chunk != null) {
                        String worldName = chunk.getWorld().getName();
                        if (worlds.contains(worldName)) checkChunk(chunk);
                    }
                }
                int count = counter.incrementAndGet();
                if (count >= checkWorldEach) {
                    counter.set(0);
                    for (String worldName : worlds) {
                        World world = Bukkit.getWorld(worldName);
                        //info("Checking world {}", world);
                        if (world == null) continue;
                        pollChunksInWorld(world);
                    }
                }
            }
        }.runTaskTimer(ARC.plugin, 60L, checkInterval);

        decayTask = new BukkitRunnable() {
            @Override
            public void run() {
                for (int i = 0; i < leafBatchSize; i++) {
                    Location leafData = leafQueue.poll();
                    if (leafData == null) return;
                    Block block = leafData.getBlock();
                    if (block.getType().isAir()) return;
                    Material type = block.getType();
                    block.breakNaturally();
                    if (playSound) block.getWorld().playSound(block.getLocation(), Sound.BLOCK_GRASS_BREAK, 1.0F, 1.0F);
                    if (playParticles) {
                        Collection<Player> nearbyPlayers = block.getWorld().getNearbyPlayers(block.getLocation(), 32);
                        ParticleManager.queue(new ParticleBuilder(Particle.BLOCK)
                                .count(2)
                                .location(block.getLocation().add(0.5, 0.5, 0.5))
                                .extra(0.1)
                                .data(type.createBlockData())
                                .offset(0.3, 0.3, 0.3)
                                .receivers(nearbyPlayers)
                        );
                    }
                }
            }
        }.runTaskTimer(ARC.plugin, 60L, decayInterval);

        tagClearTask = new BukkitRunnable() {
            @Override
            public void run() {
                while (!awaitingTagClearing.isEmpty()) {
                    Block block = awaitingTagClearing.pollFirst();
                    if (block == null) return;
                    CustomBlockData data = new CustomBlockData(block, ARC.plugin);
                    if (data.has(playerPlacedKey)) {
                        data.remove(playerPlacedKey);
                    }
                }
            }
        }.runTaskTimerAsynchronously(ARC.plugin, 60L, 1);
    }

    private static void checkChunk(Chunk chunk) {
        boolean neighborsLoaded = chunk.getWorld().isChunkLoaded(chunk.getX() + 1, chunk.getZ()) &&
                chunk.getWorld().isChunkLoaded(chunk.getX() - 1, chunk.getZ()) &&
                chunk.getWorld().isChunkLoaded(chunk.getX(), chunk.getZ() + 1) &&
                chunk.getWorld().isChunkLoaded(chunk.getX(), chunk.getZ() - 1);
        if (!neighborsLoaded) return;
        Collection<Location> leafData = leafChecker.checkChunk(chunk);
        leafQueue.addAll(leafData);
    }

    private static void pollChunksInWorld(World world) {
        if (leafQueue.size() > 10000) return;
        Collection<Player> players = world.getPlayers();
        Set<Chunk> chunks = players.stream()
                .map(Player::getLocation)
                .map(Location::getChunk)
                .flatMap(chunk -> Arrays.stream(new Chunk[]{
                        chunk,
                        world.getChunkAt(chunk.getX() + 1, chunk.getZ()),
                        world.getChunkAt(chunk.getX() - 1, chunk.getZ()),
                        world.getChunkAt(chunk.getX(), chunk.getZ() + 1),
                        world.getChunkAt(chunk.getX(), chunk.getZ() - 1),
                        world.getChunkAt(chunk.getX() + 1, chunk.getZ() + 1),
                        world.getChunkAt(chunk.getX() - 1, chunk.getZ() - 1),
                        world.getChunkAt(chunk.getX() + 1, chunk.getZ() - 1),
                        world.getChunkAt(chunk.getX() - 1, chunk.getZ() + 1)
                }))
                .collect(Collectors.toSet());
        chunkQueue.addAll(chunks);
    }

    public static void markAsPlayerPlaced(Block block) {
        if (!leafMaterials.contains(block.getType()) && !trunkMaterials.contains(block.getType())) return;
        CustomBlockData data = new CustomBlockData(block, ARC.plugin);
        data.set(playerPlacedKey, PersistentDataType.BOOLEAN, true);
    }

    public static void clearPlayerPlaced(Block block) {
        awaitingTagClearing.add(block);
    }
}

File: src/main/kotlin/ru/arc/leafdecay/LeafDecayManager.kt
package ru.arc.leafdecay

import com.destroystokyo.paper.ParticleBuilder
import com.jeff_media.customblockdata.CustomBlockData
import org.bukkit.Bukkit
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Particle
import org.bukkit.Sound
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.persistence.PersistentDataType
import ru.arc.ARC
import ru.arc.config.Config
import ru.arc.config.ConfigManager
import ru.arc.core.ScheduledTask
import ru.arc.core.repeating
import ru.arc.core.ticks
import ru.arc.util.ParticleManager
import java.util.concurrent.ConcurrentLinkedDeque
import java.util.concurrent.ConcurrentSkipListSet
import java.util.concurrent.atomic.AtomicInteger

object LeafDecayManager {

    lateinit var leafChecker: LeafChecker

    private val config: Config get() = ConfigManager.ofModule(ARC.instance.dataPath, "leafdecay.yml")

    private var checkTask: ScheduledTask? = null
    private var decayTask: ScheduledTask? = null
    private var tagClearTask: ScheduledTask? = null

    private val chunkQueue = LeafChunkQueue()
    private val leafQueue = ConcurrentLinkedDeque<Location>()
    private val awaitingTagClearing = ConcurrentLinkedDeque<Block>()

    private val playerPlacedKey by lazy { NamespacedKey(ARC.instance, "lf") }

    private var decayInterval = 5L
    private var checkInterval = 20L
    private var checkWorldEach = 1000L
    private var leafBatchSize = 100L
    private var leafMaterials: Set<Material> = emptySet()
    private var trunkMaterials: Set<Material> = emptySet()
    private var playParticles = true
    private var playSound = true
    private var worlds: Set<String> = emptySet()

    @JvmStatic
    fun init() {
        reload()
    }

    @JvmStatic
    fun reload() {
        decayInterval = config.integer("decay-interval", 5).toLong()
        checkInterval = config.integer("global-check-interval", 20).toLong()
        checkWorldEach = config.integer("world-check-interval", 1000).toLong()
        leafBatchSize = config.integer("leaf-batch-size", 100).toLong()
        leafMaterials = config.stringList("leaf-materials")
            .map { Material.valueOf(it.uppercase()) }
            .toSet()
        trunkMaterials = config.stringList("trunk-materials")
            .map { Material.valueOf(it.uppercase()) }
            .toSet()
        leafChecker = LeafChecker(config, leafMaterials, trunkMaterials)
        playParticles = config.bool("play-particles", true)
        playSound = config.bool("play-sound", true)
        worlds = config.stringList("leaf-decay-worlds").toSet()
        start()
    }

    @JvmStatic
    fun cancel() {
        stopTasks()
        chunkQueue.clear()
        leafQueue.clear()
        awaitingTagClearing.clear()
    }

    private fun stopTasks() {
        checkTask?.cancel()
        decayTask?.cancel()
        tagClearTask?.cancel()
        checkTask = null
        decayTask = null
        tagClearTask = null
    }

    private fun start() {
        stopTasks()
        val counter = AtomicInteger()

        checkTask = repeating(period = checkInterval.ticks, delay = 60.ticks) {
            var count = 0
            if (leafQueue.size <= 10000) {
                while (!chunkQueue.isEmpty()) {
                    if (count++ >= MAX_CHUNKS_PER_TICK) break
                    val chunk = chunkQueue.poll() ?: break
                    if (worlds.contains(chunk.world.name)) checkChunk(chunk)
                }
            }
            if (counter.incrementAndGet() >= checkWorldEach) {
                counter.set(0)
                for (worldName in worlds) {
                    Bukkit.getWorld(worldName)?.let { pollChunksInWorld(it) }
                }
            }
        }

        decayTask = repeating(period = decayInterval.ticks, delay = 60.ticks) {
            repeat(leafBatchSize.toInt()) {
                val location = leafQueue.poll() ?: return@repeating
                val block = location.block
                if (block.type.isAir) return@repeating
                val type = block.type
                block.breakNaturally()
                if (playSound) block.world.playSound(block.location, Sound.BLOCK_GRASS_BREAK, 1.0f, 1.0f)
                if (playParticles) {
                    val nearbyPlayers: Collection<Player> = block.world.getNearbyPlayers(block.location, 32.0)
                    ParticleManager.queue(
                        ParticleBuilder(Particle.BLOCK)
                            .count(2)
                            .location(block.location.add(0.5, 0.5, 0.5))
                            .extra(0.1)
                            .data(type.createBlockData())
                            .offset(0.3, 0.3, 0.3)
                            .receivers(nearbyPlayers)
                    )
                }
            }
        }

        tagClearTask = repeating(period = 1.ticks, delay = 60.ticks) {
            repeat(MAX_TAG_CLEARS_PER_TICK) {
                val block = awaitingTagClearing.pollFirst() ?: return@repeating
                val data = CustomBlockData(block, ARC.instance)
                if (data.has(playerPlacedKey)) {
                    data.remove(playerPlacedKey)
                }
            }
        }
    }

    private fun checkChunk(chunk: org.bukkit.Chunk) {
        val world = chunk.world
        val neighborsLoaded = world.isChunkLoaded(chunk.x + 1, chunk.z) &&
            world.isChunkLoaded(chunk.x - 1, chunk.z) &&
            world.isChunkLoaded(chunk.x, chunk.z + 1) &&
            world.isChunkLoaded(chunk.x, chunk.z - 1)
        if (!neighborsLoaded) return
        if (::leafChecker.isInitialized) leafQueue.addAll(leafChecker.checkChunk(chunk))
    }

    private fun pollChunksInWorld(world: World) {
        if (leafQueue.size > 10000) return
        world.players
            .map { it.location.chunk }
            .flatMap { chunk ->
                listOf(
                    chunk,
                    world.getChunkAt(chunk.x + 1, chunk.z),
                    world.getChunkAt(chunk.x - 1, chunk.z),
                    world.getChunkAt(chunk.x, chunk.z + 1),
                    world.getChunkAt(chunk.x, chunk.z - 1),
                    world.getChunkAt(chunk.x + 1, chunk.z + 1),
                    world.getChunkAt(chunk.x - 1, chunk.z - 1),
                    world.getChunkAt(chunk.x + 1, chunk.z - 1),
                    world.getChunkAt(chunk.x - 1, chunk.z + 1)
                )
            }
            .toSet()
            .forEach(chunkQueue::add)
    }

    @JvmStatic
    fun markAsPlayerPlaced(block: Block) {
        if (!leafMaterials.contains(block.type) && !trunkMaterials.contains(block.type)) return
        val data = CustomBlockData(block, ARC.instance)
        data.set(playerPlacedKey, PersistentDataType.BOOLEAN, true)
    }

    @JvmStatic
    fun clearPlayerPlaced(block: Block) {
        awaitingTagClearing.add(block)
    }

    private const val MAX_CHUNKS_PER_TICK = 5
    private const val MAX_TAG_CLEARS_PER_TICK = 200
}

internal class LeafChunkQueue {
    private val chunks =
        ConcurrentSkipListSet(
            compareBy<Chunk>({ it.world.uid }, { it.x }, { it.z }),
        )

    fun add(chunk: Chunk) {
        chunks.add(chunk)
    }

    fun poll(): Chunk? = chunks.pollFirst()

    fun isEmpty(): Boolean = chunks.isEmpty()

    fun size(): Int = chunks.size

    fun clear() {
        chunks.clear()
    }
}
